### PR TITLE
[Auto Parallel] Add matmul flops

### PR DESCRIPTION
**Pull request by CjhHa1**
### PR types

 New features
### PR changes

Others
### Describe

This PR adds matmul and matmul_v2 dropout elementwise_compute etc. flops computation func.

### CODE REVIEW

File: python/paddle/utils/flops.py [hunk 1]
```diff
@@ -55,6 +55,106 @@ def _dropout_flops(input_shapes, **attrs):
     return 0
 
 
+def _elementwise_flops_compute(input_shapes, **attrs):
+    input_x = input_shapes[0]
```
Comment (kuizhiqing) [inline python/paddle/utils/flops.py]: some index check would be better.

**Review by kuizhiqing — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/tests/unittests/test_profiler.py [hunk 2]
```diff
@@ -227,6 +227,44 @@ class TestFLOPSAPI(unittest.TestCase):
...
+            flops('elementwise_div', ([12, 12, 12], [2, 2, 12])) == 12 * 12 * 12
+        )
+        self.assertTrue(
+            flops(
+                'layer_norm',
+                ([12, 12, 12], [12, 12, 12], [12, 12, 12]),
+                **{'epsilon': 0.01}
+            )
+            == 12 * 12 * 12 * 8
+        )
```
Comment (kuizhiqing) [inline python/paddle/fluid/tests/unittests/test_profiler.py:298]: add unitest for dict input shape

File: python/paddle/utils/flops.py [hunk 3]
```diff
@@ -55,6 +55,147 @@ def _dropout_flops(input_shapes, **attrs):
     return 0
 
 
+def _elementwise_flops_compute(input_shapes, **attrs):
+    try:
```
Comment (kuizhiqing) [inline python/paddle/utils/flops.py]: maybe move try catch to the api function `flops`, no need do it in all funcs.

**Review by kuizhiqing — CHANGES_REQUESTED** [review]

**Comment by CjhHa1** [comment]
Get,thx

From: Zhiqing Kui ***@***.***>
Date: Tuesday, November 15, 2022 at 17:57
To: PaddlePaddle/Paddle ***@***.***>
Cc: Jianghai ***@***.***>, Author ***@***.***>
Subject: Re: [PaddlePaddle/Paddle] [Auto Parallel] Add matmul flops (PR #47816)

@kuizhiqing requested changes on this pull request.

________________________________

In python/paddle/fluid/tests/unittests/test_profiler.py<https://github.com/PaddlePaddle/Paddle/pull/47816#discussion_r1022578125>:

> +            flops('elementwise_add', ([12, 12, 12], [2, 2, 12])) == 12 * 12 * 12

+        )

+        self.assertTrue(

+            flops('elementwise_mul', ([12, 12, 12], [2, 2, 12])) == 12 * 12 * 12

+        )

+        self.assertTrue(

+            flops('elementwise_div', ([12, 12, 12], [2, 2, 12])) == 12 * 12 * 12

+        )

+        self.assertTrue(

+            flops(

+                'layer_norm',

+                ([12, 12, 12], [12, 12, 12], [12, 12, 12]),

+                **{'epsilon': 0.01}

+            )

+            == 12 * 12 * 12 * 8

+        )

add unitest for dict input shape

________________________________

In python/paddle/utils/flops.py<https://github.com/PaddlePaddle/Paddle/pull/47816#discussion_r1022579053>:

> @@ -55,6 +55,147 @@ def _dropout_flops(input_shapes, **attrs):

     return 0





+def _elementwise_flops_compute(input_shapes, **attrs):

+    try:

maybe move try catch to the api function flops, no need do it in all funcs.

File: python/paddle/utils/flops.py [hunk 4]
```diff
@@ -51,10 +55,207 @@ def register(func):
...
+        in_x = input_x[dim_x - 1 - i] if i < dim_x else 1
+        in_y = input_y[dim_y - 1 - i] if i < dim_y else 1
+        output.append(max(in_x, in_y))
+    return prod(output)
+
+
+@register_flops("elementwise_add")
+def _elementwise_add_flops(input_shapes, attrs):
+    """
+    flops computation for elementwise_add op
```
Comment (aoyulong) [inline python/paddle/utils/flops.py]: Please write the comment in a standard way.

File: python/paddle/utils/flops.py [hunk 5]
```diff
@@ -51,10 +55,207 @@ def register(func):
...
+def _reshape2_flops(input_shapes, attrs):
+    """
+    flops computation for reshape2 op
+    equation: flops = 0
+    """
+    return 0
+
+
+@register_flops("soft_max")
+def _soft_max_flops(input_shapes, attrs):
```
Comment (aoyulong) [inline python/paddle/utils/flops.py]: Please rename `_soft_max` to `softmax`.

**Review by kuizhiqing — CHANGES_REQUESTED** [review]
/hold /do-not-merge

**Review by aoyulong — APPROVED** [review]
LGTM

**Review by JZ-LIANG — APPROVED** [review]
LGTM

**Pull request merged**